// backed/backed.js
var Backed = (function () {
   'use strict';

   /**
    * @mixin Backed
    * @module utils
    * @export loadScript
    *
    * defer handles loading after the document is parsed, async loads while parsing
    *
    * @param {string} src link/path to the script to load
    * @param {string} method default: 'async',  options: `defer, async, ''`
    * @param {string} type default: undefined,  options: `module, utf-8, ...`
    * @return {object} merge result
    */
    var loadScript = (src, method = 'async', type) => {
      return new Promise((resolve, reject) => {
        let script = document.createElement('script');
        script.setAttribute(method, '');
        if (type) script.setAttribute('type', type);
        script.onload = result => {
          resolve(result);
        };
        script.onerror = error => {
          reject(error);
        };
        script.src = src;
        document.body.appendChild(script);
      });
    }

   /**
    * @mixin backed
    * @module utils
    * @export fireEvent
    *
    * @param {string} type Name of the event
    * @param {HTMLElement} target context
    * @param {string|boolean|number|object|array} detail
    */
   var fireEvent = (type='', detail=null, target=HTMLElement) => {
     target.dispatchEvent(new CustomEvent(type, {detail: detail}));
   };

   /**
    * @mixin Backed
    * @module utils
    * @export toJsProp
    *
    * some-prop -> someProp
    *
    * @param {string} string The content to convert
    * @return {string} string
    */
   var toJsProp = string => {
     let parts = string.split('-');
     if (parts.length > 1) {
       var upper = parts[1].charAt(0).toUpperCase();
       string = parts[0] + upper + parts[1].slice(1).toLowerCase();
     }
     return string;
   };

   let sheduled = false;
   const afterRenderQue = [];
   const beforeRenderQue = [];

   const callMethod = array => {
     const context = array[0];
     const callback = array[1];
     const args = array[2];
     try {
       callback.apply(context, args);
     } catch(e) {
       setTimeout(() => {
         throw e;
       });
     }
   };

   const flushQue = que => {
     while (que.length) {
       callMethod(que.shift);
     }
   };

   const runQue = que => {
     for (let i=0, l=que.length; i < l; i++) {
       callMethod(que.shift());
     }
     sheduled = false;
   };

   const shedule = () => {
     sheduled = true;
     requestAnimationFrame(() => {
       flushQue(beforeRenderQue);
       setTimeout(() => {
         runQue(afterRenderQue);
       });
     });
   };

   (() => {
     window.RenderStatus = window.RenderStatus || {
       afterRender: (context, callback, args) => {
         if (!sheduled) {
           shedule();
         }
         afterRenderQue.push([context, callback, args]);
       },
       beforeRender: (context, callback, args) => {
         if (!sheduled) {
           shedule();
         }
         beforeRenderQue.push([context, callback, args]);
       }
     };
   })();

   class PubSub {

     /**
      * Creates handlers
      */
     constructor() {
       this.handlers = [];
     }

     /**
      * @param {String} event
      * @param {Method} handler
      * @param {HTMLElement} context
      */
     subscribe(event, handler, context) {
       if (typeof context === 'undefined') {
         context = handler;
       }
       this.handlers.push({event: event, handler: handler.bind(context)});
     }

     /**
      * @param {String} event
      * @param {String|Number|Boolean|Object|Array} change
      */
     publish(event, change) {
       for (let i = 0; i < this.handlers.length; i++) {
         if (this.handlers[i].event === event) {
           let oldValue = this.handlers[i].oldValue;
           // dirty checking value, ensures that we don't create a loop
           if (oldValue !== change.value) {
             this.handlers[i].handler(change, this.handlers[i].oldValue);
             this.handlers[i].oldValue = change.value;
           }
         }
       }
     }
   }

   var PubSub$1 = window.PubSub ? window.PubSub : window.PubSub = new PubSub();

   window.registeredElements = window.registeredElements || [];

   const handleProperties = (target, properties) => {
     if (properties) {
       for (let entry of Object.entries(properties)) {
         handleProperty(target, entry[0], entry[1]);
         // TODO: are we ignoring stuff ...?
         // check if attribute has value else pass default property value
         target[entry[0]] = target.hasAttribute(entry[0]) ? target.getAttribute(entry[0]) : entry[1].value;
       }
     }
   };

   const handleProperty = (obj, property, {observer, strict, global, reflect, renderer, value }) => {

     if (Boolean(observer || global) && _needsObserverSetup(obj, property)) {
       // Ensure we don't do duplicate work
       obj.observedProperties.push(property);

       // subscribe only when a callback is defined, all other global options are still available ...
       if (global && obj[observer]) {
         // Warning, global observers don't work the same like bindings, each observer has it's namespace created like global.name,
         // so whenever another element has an global observer for name, they will subscribe to the same publisher !
         // TODO: Add local binding & improve global observers
         // {{name}} for normal bindings & {{global::name}} for global bindings(observers) (like Polymer does)
         // this means we need to build a system that keeps track of each component it's bindings &
         // values should be set as property, so we know if a value needs to be set on attribute, rerender template, etc ..
         PubSub$1.subscribe(`global.${property}`, obj[observer].bind(obj));
       }
       setupObserver(obj, property, observer, {strict, global, reflect, renderer: renderer});
     } else if (!Boolean(observer || global) && Boolean(reflect || renderer)) {
       setupObserver(obj, property, observer, {strict, global, reflect, renderer: renderer});
     }
   };

   const _needsObserverSetup = (obj, property) => {
     if (!obj.observedProperties) {
       obj.observedProperties = [];
     }
     if (obj.observedProperties[property]) {
       console.warn(
         'observer::ignoring duplicate property observer ' + property
       );
       return false;
     } else {
       return true;
     }
   };

   const forObservers = (target, observers, isGlobal=false) => {
     for (let observe of observers) {
       let parts = observe.split(/\(|\)/g);
       let fn = parts[0];
       parts = parts.slice(1);
       for (let property of parts) {
         if (property.length) {
           handleProperty(target, property, fn, {
             strict: false,
             global: isGlobal
           });
         }
       }
     }
   };

   /**
    * Runs a method on target whenever given property changes
    *
    * example:
    * change(change) {
    *  change.property // name of the property
    *  change.value // value of the property
    * }
    *
    * @arg {object} obj target
    * @arg {string} property name
    * @arg {boolean} strict
    * @arg {method} fn The method to run on change
    */
   const setupObserver = (obj, property, fn, {strict, global, reflect, renderer}) => {
     if (renderer && !obj.shadowRoot) {
       obj.attachShadow({mode: 'open'});
     }
     Object.defineProperty(obj, property, {
       set(value) {
         if (value === undefined) {
           return
         }
         if (this[`_${property}`] === value) {
           return;
         }
         this[`_${property}`] = value;
         let data = {
           property: property,
           value: value
         };
         if (reflect) {
           if (value) this.setAttribute(property, String(value));
           else this.removeAttribute(property);
         }
         if (renderer) {
             // adds support for multiple renderers
             const object = {};
             object[property] = value;
             this.render(object, this[renderer]);
         }
         if (global) {
           data.instance = this;
           PubSub$1.publish(`global.${property}`, data);
         } else if(fn) {
           if (this[fn]) {
             this[fn](data);
           } else {
             console.warn(`observer undefined::${fn} is not a function`);
           }
         }
       },
       get() {
         return this[`_${property}`];
       },
       configurable: strict ? false : true
     });
   };


   const handleObservers = (target, observers=[], globalObservers=[]) => {
     if (!observers && !globalObservers) {
       return;
     }
     forObservers(target, observers);
   };

   const handleListeners = target => {
     const attributes = target.attributes;
     for (const attribute of attributes) {
       if (String(attribute.name).includes('on-')) {
         const fn = attribute.value;
         const name = attribute.name.replace('on-', '');
         target.addEventListener(String(name), event => {
           target = event.path[0];
           while (!target.host) {
             target = target.parentNode;
           }
           if (target.host[fn]) {
             target.host[fn](event);
           }
         });
       }
     }
   };

   const ready = target => {
     requestAnimationFrame(() => {
       if (target.ready) target.ready();
     });
   };

   const constructorCallback = (target=HTMLElement, klass=Function, hasWindow=false) => {

     target.fireEvent = target.fireEvent || fireEvent.bind(target);
     target.toJsProp = target.toJsProp || toJsProp.bind(target);
     target.loadScript = target.loadScript || loadScript.bind(target);


     // setup properties

     if (!target.registered && target.created) target.created();

     // let backed know the element is registered
     target.registered = true;
   };

   const connectedCallback = (target=HTMLElement, klass=Function) => {
     if (target.connected) target.connected();
     handleProperties(target, klass.properties);
     handleObservers(target, klass.observers, klass.globalObservers);
     // setup listeners
     handleListeners(target);
     // notify everything is ready
     ready(target);
   };

   const shouldRegister = (name, klass) => {
     if (window.registeredElements.indexOf(name) === -1) {
       window.registeredElements.push(name);
       return true;
     }
     return false;
   };

   var base = {
     ready: ready,
     connectedCallback: connectedCallback,
     constructorCallback: constructorCallback,
     shouldRegister: shouldRegister
   }

   /**
    * Add space between camelCase text.
    */
   var unCamelCase = (string) => {
     string = string.replace(/([a-z\xE0-\xFF])([A-Z\xC0\xDF])/g, '$1 $2');
     string = string.toLowerCase();
     return string;
   };

   /**
   * Replaces all accented chars with regular ones
   */
   var replaceAccents = (string) => {
     // verifies if the String has accents and replace them
     if (string.search(/[\xC0-\xFF]/g) > -1) {
         string = string
                 .replace(/[\xC0-\xC5]/g, 'A')
                 .replace(/[\xC6]/g, 'AE')
                 .replace(/[\xC7]/g, 'C')
                 .replace(/[\xC8-\xCB]/g, 'E')
                 .replace(/[\xCC-\xCF]/g, 'I')
                 .replace(/[\xD0]/g, 'D')
                 .replace(/[\xD1]/g, 'N')
                 .replace(/[\xD2-\xD6\xD8]/g, 'O')
                 .replace(/[\xD9-\xDC]/g, 'U')
                 .replace(/[\xDD]/g, 'Y')
                 .replace(/[\xDE]/g, 'P')
                 .replace(/[\xE0-\xE5]/g, 'a')
                 .replace(/[\xE6]/g, 'ae')
                 .replace(/[\xE7]/g, 'c')
                 .replace(/[\xE8-\xEB]/g, 'e')
                 .replace(/[\xEC-\xEF]/g, 'i')
                 .replace(/[\xF1]/g, 'n')
                 .replace(/[\xF2-\xF6\xF8]/g, 'o')
                 .replace(/[\xF9-\xFC]/g, 'u')
                 .replace(/[\xFE]/g, 'p')
                 .replace(/[\xFD\xFF]/g, 'y');
     }

     return string;
   };

   var removeNonWord = (string) => string.replace(/[^0-9a-zA-Z\xC0-\xFF \-]/g, '');

   const WHITE_SPACES = [
       ' ', '\n', '\r', '\t', '\f', '\v', '\u00A0', '\u1680', '\u180E',
       '\u2000', '\u2001', '\u2002', '\u2003', '\u2004', '\u2005', '\u2006',
       '\u2007', '\u2008', '\u2009', '\u200A', '\u2028', '\u2029', '\u202F',
       '\u205F', '\u3000'
   ];

   /**
   * Remove chars from beginning of string.
   */
   var ltrim = (string, chars) => {
     chars = chars || WHITE_SPACES;

     let start = 0,
         len = string.length,
         charLen = chars.length,
         found = true,
         i, c;

     while (found && start < len) {
         found = false;
         i = -1;
         c = string.charAt(start);

         while (++i < charLen) {
             if (c === chars[i]) {
                 found = true;
                 start++;
                 break;
             }
         }
     }

     return (start >= len) ? '' : string.substr(start, len);
   };

   /**
   * Remove chars from end of string.
   */
   var rtrim = (string, chars) => {
     chars = chars || WHITE_SPACES;

     var end = string.length - 1,
         charLen = chars.length,
         found = true,
         i, c;

     while (found && end >= 0) {
         found = false;
         i = -1;
         c = string.charAt(end);

         while (++i < charLen) {
             if (c === chars[i]) {
                 found = true;
                 end--;
                 break;
             }
         }
     }

     return (end >= 0) ? string.substring(0, end + 1) : '';
   }

   /**
    * Remove white-spaces from beginning and end of string.
    */
   var trim = (string, chars) => {
     chars = chars || WHITE_SPACES;
     return ltrim(rtrim(string, chars), chars);
   }

   /**
    * Convert to lower case, remove accents, remove non-word chars and
    * replace spaces with the specified delimeter.
    * Does not split camelCase text.
    */
   var slugify = (string, delimeter) => {
     if (delimeter == null) {
         delimeter = "-";
     }

     string = replaceAccents(string);
     string = removeNonWord(string);
     string = trim(string) //should come after removeNonWord
             .replace(/ +/g, delimeter) //replace spaces with delimeter
             .toLowerCase();
     return string;
   };

   /**
   * Replaces spaces with hyphens, split camelCase text, remove non-word chars, remove accents and convert to lower case.
   */
   var hyphenate = string => {
     string = unCamelCase(string);
     return slugify(string, "-");
   }

   const shouldRegister$1 = name => {
     return customElements.get(name) ? false : true;
   };

   /**
    *
    * @module backed
    * @param {class} _class
    */
   var backed = _class => {
     let klass;
     // get the tagName or try to make one with class.name
     let name = _class.is || hyphenate(_class.name);
     klass = class extends _class {
       static get observedAttributes() {
         return Object.entries(_class.properties).map(entry => {if (entry[1].reflect) {return entry[0]} else return null});
       }
       constructor() {
         super();
         base.constructorCallback(this, _class);
       }
       connectedCallback() {
         base.connectedCallback(this, _class);
       }
       attributeChangedCallback(name, oldValue, newValue) {
         this[name] = newValue;
       }
     };
     if (shouldRegister$1(name, klass)) {
       customElements.define(name, klass);
     }};

   return backed;

}());
